symbolic-base/src/ZkFold/Symbolic/Data/Int.hs

```diff
@@ -0,0 +1,216 @@
+isNegative (Int u) =
+    withDict (plusNat @n @1) $
+        withNumberOfRegisters @(n+1) @r @(BaseField c) $
+            withGetRegisterSize @(n+1) @r @(BaseField c) $
+                withCeilRegSize @(GetRegisterSize (BaseField c) (n+1) r) @OrdWord $
+                    u >= fromConstant ((2 :: Natural) ^ (value @n))
+
+isNotNegative :: forall n r c.
+    ( Symbolic c, KnownNat n, KnownRegisterSize r
+    ) => Int n r c -> Bool c
+isNotNegative i = not (isNegative i)
+
+abs :: forall c n r.
+    ( Symbolic c, KnownNat n, KnownRegisterSize r
+    )  => Int n r c -> Int n r c
+abs i = withDict (plusNat @n @1) $
+            withNumberOfRegisters @(n+1) @r @(BaseField c) $
+                bool i (negate i) (isNegative i)
+
+instance (Symbolic c, KnownNat (n+1), KnownNat n, KnownRegisterSize r) => Scale Natural (Int n r c)
+instance (Symbolic c, KnownNat (n+1), KnownNat n, KnownRegisterSize r) => Scale Integer (Int n r c)
+instance (Symbolic c, KnownNat (n+1), KnownNat n, KnownRegisterSize r) => MultiplicativeMonoid (Int n r c) where
+    one = fromConstant (1 :: Natural)
+
+instance (Symbolic c, KnownNat (n+1), KnownNat n, KnownRegisterSize r) => AdditiveMonoid (Int n r c) where
+    zero = fromConstant (0:: Natural)
+
+instance (Symbolic c, KnownNat (n+1), KnownNat n, KnownRegisterSize r) => Semiring (Int n r c)
+instance (Symbolic c, KnownNat (n+1), KnownRegisterSize r) => Arbitrary (Int n r c) where
+    arbitrary = Int Haskell.<$> arbitrary
+
+instance (Symbolic c, KnownNat n, k ~ n+1, KnownNat k, KnownRegisterSize r) => Iso (Int n r c) (UInt k r c) where
+    from (Int u) = u
+
+instance (Symbolic c, KnownNat n, k ~ n+1, KnownNat k, KnownRegisterSize r) => Iso (UInt k r c) (Int n r c) where
+    from = Int
+
+-- -- --------------------------------------------------------------------------------
+instance (Symbolic c, KnownNat (n+1), KnownRegisterSize r) => AdditiveSemigroup (Int n r c) where
+    Int u1 + Int u2 = Int (u1+u2)
+
+instance (Symbolic c, KnownNat (n+1), KnownNat n, KnownRegisterSize r) => AdditiveGroup (Int n r c) where
+    Int u1 - Int u2 = Int (u1-u2)
+    negate (Int u) = Int (negate u)
+
+instance (Symbolic c, KnownNat (n+1), KnownRegisterSize r) => MultiplicativeSemigroup (Int n r c) where
+    Int x * Int y = Int $ x * y
+
+instance ( Symbolic c, KnownNat (n+1), KnownNat n, KnownRegisterSize r
+         , KnownRegisters c (n+1) r
+         ) => SemiEuclidean (Int n r c) where
+  divMod i1 i2 = ifThenElse (isNegative i1 && isNegative i2) dm_mm ite_tf
+    where
+        (Int u1, Int u2) = (abs i1, abs i2)
+        (d, m) = withGetRegisterSize @(n+1) @r @(BaseField c) $
+                    withCeilRegSize @(GetRegisterSize (BaseField c) (n+1) r) @OrdWord $
+                        divMod u1 u2
+
+
+        ite_tf = ifThenElse (isNegative i1 && not (isNegative i2)) dm_mp ite_pm
+        ite_pm = ifThenElse (not (isNegative i1) && isNegative i2) dm_pm dm_pp
+
+        dm_mm = (Int d, negate (Int m))
```

Comment:
Modulus is always positive by definition, I think we should stick to the definition